# Lab 01: MỐI QUAN HỆ CỦA DỮ LIỆU
# Môn: Trực Quan Hóa Dữ Liệu

## Danh sách thành viên:

---
**Họ và tên:** Nguyễn Tất Trường

MSSV: 19127082

---
**Họ và tên:** Nguyễn Hữu Tuấn

**MSSV:** 19127083

---
**Họ và tên:** Chung Thế Thọ

**MSSV:** 19127562

---
**Họ và tên:** Trần Quốc Việt

**MSSV:** 19127630

# Import các thư viện cần thiết

In [1]:
# Import thư viện cần dùng
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt

# Đường dẫn đến các thư mục
datapath = 'dataset/'

# Thu thập dữ liệu

Trong đồ án này, ta sẽ sử dụng tập dữ liệu về đại dịch **Coronavirus** thu thập từ trang **Worldometer** (www.worldometers.info/coronavirus/). Dữ liệu được thu thập từ ngày *20-02-2022* đến ngày *10-03-2022* và không liên tục, nghĩa là có vài ngày chưa thu thập dữ liệu. Theo như mô tả trên Worldometer thì dữ liệu của trang này cung cấp cho ta thông tin về sự tác động của **Coronavirus** lên **225 đất nước và vùng lãnh thổ** khác nhau qua từng ngày.

Về cách thức thu thập dữ liệu thì có thể xem chi tiết ở notebook `\dataset\scraping_worldometers.ipynb`

Link của tập dữ liệu đã thu thập: ...

---

# Khám phá dữ liệu

## Đọc dữ liệu của ngày đầu tiên từ file csv vào dataframe

Dữ liệu dạng số lưu trữ trong file được phân cách phần ngàn với nhau bằng dấu phẩy (ví dụ 1,000,000), vì vậy ta truyền vào tham số **thousand=','** khi đọc dữ liệu để cho thư viện pandas tự động chuyển các số này về dạng số thông thường

In [2]:
filename = '2022-2-20.csv'
filepath = datapath + filename
first_day_df = pd.read_csv(filepath, thousands=',')
first_day_df.head(10)

#  Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0  NaN           Asia    112080501   521505.0     1333376.0      1766.0   
1  NaN  North America     94158432    57642.0     1384321.0       979.0   
2  NaN  South America     53321618    91557.0     1248687.0       984.0   
3  NaN         Europe    150719217   660784.0     1681452.0      2030.0   
4  NaN         Africa     11442875     8624.0      247689.0       194.0   
5  NaN        Oceania      3284142    20201.0        7457.0        49.0   
6  NaN            NaN          721        NaN          15.0         NaN   
7  NaN          World    425007506  1360313.0     5902997.0      6002.0   
8  1.0          China       107707      195.0        4636.0         NaN   
9  2.0            USA     80117484    28647.0      959863.0       431.0   

   Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0      104076728.0       498347.0     6670397.0            30563.0   
1       63866758.0       164163.0    28907353.0            18121.0   
2       46304681.0       148298.0     5768250.0            13954.0   
3      122959179.0      1165527.0    26078586.0            16595.0   
4       10454264.0        15330.0      740922.0             1984.0   
5        2997343.0        21041.0      279342.0              263.0   
6            706.0            NaN           0.0                0.0   
7      350659659.0      2012706.0    68444850.0            81480.0   
8         101449.0           55.0        1622.0                7.0   
9       51675064.0       129104.0    27482557.0            11668.0   

   Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop    Population  \
0               NaN            NaN          NaN           NaN           NaN   
1               NaN            NaN          NaN           NaN           NaN   
2               NaN            NaN          NaN           NaN           NaN   
3               NaN            NaN          NaN           NaN           NaN   
4               NaN            NaN          NaN           NaN           NaN   
5               NaN            NaN          NaN           NaN           NaN   
6               NaN            NaN          NaN           NaN           NaN   
7           54525.0          757.3          NaN           NaN           NaN   
8              75.0            3.0  160000000.0      111163.0  1.439324e+09   
9          239741.0         2872.0  938796430.0     2809224.0  3.341836e+08   

              Region  
0               Asia  
1      North America  
2      South America  
3             Europe  
4             Africa  
5  Australia/Oceania  
6                NaN  
7                All  
8               Asia  
9      North America

## Mỗi cột có ý nghĩa gì?

- **#**: Số thứ tự
- **Country,Other**: Tên đất nước/vùng lãnh thổ
- **Total Cases**: Tổng số ca nhiễm
- **New Cases**: Số ca mới trong ngày
- **Total Deaths**: Tổng số ca tử vong
- **New Deaths**: Số ca tử vong trong ngày
- **Total Recovered**: Tổng số ca đã khỏi bệnh
- **New Recovered**: Số ca khỏi bệnh trong ngày
- **Active Cases**: Số ca đang còn bệnh
- **Serious, Critical**: Số ca nghiêm trọng
- **Tot Cases/1M pop**: Tổng ca nhiễm/triệu người
- **Deaths/1M pop**: Tổng ca tử vong/triệu người
- **Total Tests**: Tổng số lần test
- **Tests/1M pop**: Số test/triệu người
- **Population**: Dân số của đất nước/vùng lãnh thổ
- **Region**: Khu vực của đất nước/vùng lãnh thổ

## Số dòng, số cột của tập dữ liệu

In [3]:
num_rows, num_cols = first_day_df.shape
print('Số dòng:', num_rows)
print('Số cột:', num_cols)

Số dòng: 234
Số cột: 16


**Nhận xét:**

Số dòng thực tế nhiều hơn so với số dòng của mô tả trên Worldometers (234 so với 225). Nguyên do của việc này ta có thể thấy ở đầu tập dữ liệu:
- 7 dòng tổng hợp số liệu theo Region/World
- 1 dòng không có ý nghĩa (dòng thứ 7)
Ngoài ra có thể mô tả trên trang Worldometer chưa được cập nhật, bằng chứng là khi ta xem dòng cuối cùng của tập dữ liệu, ta thấy số thứ tự thông qua cột **#** không phải là 225.

In [4]:
first_day_df.tail(1)

# Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
233  226.0    Micronesia            1        NaN           NaN         NaN   

     Total Recovered  New Recovered  Active Cases  Serious, Critical  \
233              1.0            NaN           0.0                NaN   

     Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
233               9.0            NaN          NaN           NaN    116999.0   

                Region  
233  Australia/Oceania

## Mỗi dòng có ý nghĩa gì? Có vấn đề các dòng có ý nghĩa khác nhau không?

Mỗi dòng trong tập dữ liệu cho ta biết thông tin về tình hình các ca nhiễm, ca khỏi Covid-19 của một đất nước/vùng lãnh thổ cụ thể. Ngoài trừ dòng thứ 7 không mang ý nghĩa mà ta đã đề cập trên, các dòng còn lại có vẻ như không có gì bất ổn. Vì vậy ta sẽ tiến hành xóa đi dòng không có ý nghĩa để thuận tiện cho xử lý sau này.

In [5]:
first_day_df = first_day_df.drop(6)

## Dữ liệu có bị lặp không

In [6]:
print("Duplicated count:", np.sum(first_day_df.duplicated()))

Duplicated count: 0


## Kiểu dữ liệu của từng cột

In [7]:
first_day_df.dtypes

#                    float64
Country,Other         object
Total Cases            int64
New Cases            float64
Total Deaths         float64
New Deaths           float64
Total Recovered      float64
New Recovered        float64
Active Cases         float64
Serious, Critical    float64
Tot Cases/1M pop     float64
Deaths/1M pop        float64
Total Tests          float64
Tests/1M pop         float64
Population           float64
Region                object
dtype: object

**Nhận xét:**

Có vẻ kiểu dữ liệu của các cột không có gì bất ổn. Để đảm bảo ta đi xem các cột mang kiểu *object* có chỉ chứa các dữ liệu dạng chuỗi hay không.

## Phân tích rõ hơn các cột có kiểu object

In [8]:
def open_object_dtype(s):
    return set(s.apply(type))
    
print('Country,Other:', open_object_dtype(first_day_df['Country,Other']))
print('Region:', open_object_dtype(first_day_df['Region']))

Country,Other: {<class 'str'>}
Region: {<class 'str'>, <class 'float'>}


**Nhận xét:**

Kiểu dữ liệu ở cột **Country,Other** đã đúng như ta mong đợi, tuy nhiên ở cột **Region** thì theo kết quả cho ta thấy, có các dòng nào đó mang kiểu dữ liệu float. Ta sẽ xem thử dữ liệu của các dòng này.

In [9]:
first_day_df[first_day_df['Region'].apply(type) == float]

#     Country,Other  Total Cases  New Cases  Total Deaths  \
219  212.0  Diamond Princess          712        NaN          13.0   
229  222.0        MS Zaandam            9        NaN           2.0   

     New Deaths  Total Recovered  New Recovered  Active Cases  \
219         NaN            699.0            NaN           0.0   
229         NaN              7.0            NaN           0.0   

     Serious, Critical  Tot Cases/1M pop  Deaths/1M pop  Total Tests  \
219                NaN               NaN            NaN          NaN   
229                NaN               NaN            NaN          NaN   

     Tests/1M pop  Population Region  
219           NaN         NaN    NaN  
229           NaN         NaN    NaN

Hai dòng kết quả trả về là dữ liệu của **Diamond Princess** và **MS Zaandam**. Nếu như theo dõi báo đài, tin tức thì ta sẽ biết đây là tên của hai con du thuyền du lịch và vì là du thuyền nên chúng không có **Region** xác định.

## Các cột có bị thiếu dữ liệu không?

In [10]:
first_day_df.isnull().any()

#                     True
Country,Other        False
Total Cases          False
New Cases             True
Total Deaths          True
New Deaths            True
Total Recovered       True
New Recovered         True
Active Cases          True
Serious, Critical     True
Tot Cases/1M pop      True
Deaths/1M pop         True
Total Tests           True
Tests/1M pop          True
Population            True
Region                True
dtype: bool

**Nhận xét:**
    
Chỉ có 2 cột **Country,Other** và **Total Cases** là đầy đủ còn lại tất cả các cột đều bị thiếu dữ liệu.

---

# Tiền xử lý dữ liệu

Thông qua việc khám phá dữ liệu, ta có các vấn đề cần tiền xử lý sau:
- Loại bỏ cột **#**
- Điền dữ liệu bị thiếu ở các cột
- Thêm cột **Date**: ngày thu thập dữ liệu

## Loại bỏ cột #

Ý nghĩa của cột này là số thứ tự mà ta đã có index trong dataframe của pandas nên ta không cần dùng đến cột này nữa để tiết kiệm không gian lưu trữ.

In [11]:
first_day_df = first_day_df.drop('#', axis=1)
first_day_df.head()

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0           Asia    112080501   521505.0     1333376.0      1766.0   
1  North America     94158432    57642.0     1384321.0       979.0   
2  South America     53321618    91557.0     1248687.0       984.0   
3         Europe    150719217   660784.0     1681452.0      2030.0   
4         Africa     11442875     8624.0      247689.0       194.0   

   Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0      104076728.0       498347.0     6670397.0            30563.0   
1       63866758.0       164163.0    28907353.0            18121.0   
2       46304681.0       148298.0     5768250.0            13954.0   
3      122959179.0      1165527.0    26078586.0            16595.0   
4       10454264.0        15330.0      740922.0             1984.0   

   Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0               NaN            NaN          NaN           NaN         NaN   
1               NaN            NaN          NaN           NaN         NaN   
2               NaN            NaN          NaN           NaN         NaN   
3               NaN            NaN          NaN           NaN         NaN   
4               NaN            NaN          NaN           NaN         NaN   

          Region  
0           Asia  
1  North America  
2  South America  
3         Europe  
4         Africa

## Điền dữ liệu bị thiếu ở các cột

### Điền giá trị 0

Các cột có thể điền giá trị 0 là: **"New Cases", "Total Deaths", "New Deaths", "Total Recovered", "New Recovered", "Active Cases", "Serious, Critical", "Total Tests"**. Việc điền 0 vào các cột này là có thể chấp nhận được vì có thể một số quốc gia/vùng lãnh thổ không công bố số liệu nữa hoặc chỉ công bố sau một thời gian dài dẫn đến việc ta không có được số liệu cập nhật hằng ngày ở các quốc gia/vùng lãnh thổ này. Vì thế thay vì đoán giá trị của các cột này, tốt nhất ta cho chúng giá trị 0.

In [12]:
fill_zero_cols = ["New Cases", "Total Deaths", "New Deaths", "Total Recovered",
                  "New Recovered", "Active Cases", "Serious, Critical", "Total Tests"]
first_day_df[fill_zero_cols] = first_day_df[fill_zero_cols].fillna(0)

### Điền giá trị thiếu cho cột Population

In [13]:
first_day_df[first_day_df['Population'].isnull()]

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0                Asia    112080501   521505.0     1333376.0      1766.0   
1       North America     94158432    57642.0     1384321.0       979.0   
2       South America     53321618    91557.0     1248687.0       984.0   
3              Europe    150719217   660784.0     1681452.0      2030.0   
4              Africa     11442875     8624.0      247689.0       194.0   
5             Oceania      3284142    20201.0        7457.0        49.0   
7               World    425007506  1360313.0     5902997.0      6002.0   
219  Diamond Princess          712        0.0          13.0         0.0   
229        MS Zaandam            9        0.0           2.0         0.0   

     Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0        104076728.0       498347.0     6670397.0            30563.0   
1         63866758.0       164163.0    28907353.0            18121.0   
2         46304681.0       148298.0     5768250.0            13954.0   
3        122959179.0      1165527.0    26078586.0            16595.0   
4         10454264.0        15330.0      740922.0             1984.0   
5          2997343.0        21041.0      279342.0              263.0   
7        350659659.0      2012706.0    68444850.0            81480.0   
219            699.0            0.0           0.0                0.0   
229              7.0            0.0           0.0                0.0   

     Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop  Population  \
0                 NaN            NaN          0.0           NaN         NaN   
1                 NaN            NaN          0.0           NaN         NaN   
2                 NaN            NaN          0.0           NaN         NaN   
3                 NaN            NaN          0.0           NaN         NaN   
4                 NaN            NaN          0.0           NaN         NaN   
5                 NaN            NaN          0.0           NaN         NaN   
7             54525.0          757.3          0.0           NaN         NaN   
219               NaN            NaN          0.0           NaN         NaN   
229               NaN            NaN          0.0           NaN         NaN   

                Region  
0                 Asia  
1        North America  
2        South America  
3               Europe  
4               Africa  
5    Australia/Oceania  
7                  All  
219                NaN  
229                NaN

Các dòng thiếu dữ liệu ở cột này là các dòng tổng hợp số liệu theo **Region/World** và 2 con du thuyền ta đã đề cập ở trên. Đối với các dòng tổng hợp thì ta sẽ gom nhóm các quốc gia/vùng lãnh thổ theo **Region**, tính tổng theo cột **Population** rồi lấy kết quả để điền, còn đối với 2 chiếc du thuyền thi ta sẽ điền 0.

In [14]:
# Hai chiếc du thuyền
two_cruises = first_day_df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
first_day_df.loc[two_cruises, 'Population'] = 0

# Gôm nhóm theo region rồi tỉnh tổng theo cột Population
region_pop = first_day_df.groupby('Region')['Population'].sum()
region_pop['All'] = region_pop.sum() # World

# Định nghĩa hàm trả Population khi truyền vào region
def region_population(region_pop, region):
    return region_pop[region]

# Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
first_day_df.loc[0:7, 'Population'] = first_day_df.head(7).apply(lambda row: region_population(region_pop, row['Region']), axis=1)

first_day_df.head(10)

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0            Asia    112080501   521505.0     1333376.0      1766.0   
1   North America     94158432    57642.0     1384321.0       979.0   
2   South America     53321618    91557.0     1248687.0       984.0   
3          Europe    150719217   660784.0     1681452.0      2030.0   
4          Africa     11442875     8624.0      247689.0       194.0   
5         Oceania      3284142    20201.0        7457.0        49.0   
7           World    425007506  1360313.0     5902997.0      6002.0   
8           China       107707      195.0        4636.0         0.0   
9             USA     80117484    28647.0      959863.0       431.0   
10          India     42838524    16051.0      512141.0       206.0   

    Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0       104076728.0       498347.0     6670397.0            30563.0   
1        63866758.0       164163.0    28907353.0            18121.0   
2        46304681.0       148298.0     5768250.0            13954.0   
3       122959179.0      1165527.0    26078586.0            16595.0   
4        10454264.0        15330.0      740922.0             1984.0   
5         2997343.0        21041.0      279342.0              263.0   
7       350659659.0      2012706.0    68444850.0            81480.0   
8          101449.0           55.0        1622.0                7.0   
9        51675064.0       129104.0    27482557.0            11668.0   
10       42124284.0        37901.0      202099.0             8944.0   

    Tot Cases/1M pop  Deaths/1M pop  Total Tests  Tests/1M pop    Population  \
0                NaN            NaN          0.0           NaN  4.664923e+09   
1                NaN            NaN          0.0           NaN  5.966738e+08   
2                NaN            NaN          0.0           NaN  4.365839e+08   
3                NaN            NaN          0.0           NaN  7.483700e+08   
4                NaN            NaN          0.0           NaN  1.393144e+09   
5                NaN            NaN          0.0           NaN  4.337959e+07   
7            54525.0          757.3          0.0           NaN  7.883074e+09   
8               75.0            3.0  160000000.0      111163.0  1.439324e+09   
9           239741.0         2872.0  938796430.0     2809224.0  3.341836e+08   
10           30550.0          365.0  760146333.0      542096.0  1.402236e+09   

               Region  
0                Asia  
1       North America  
2       South America  
3              Europe  
4              Africa  
5   Australia/Oceania  
7                 All  
8                Asia  
9       North America  
10               Asia

### Điền lại giá trị cột Total Tests cho các dòng tổng hợp theo Region/World

Tương tự ta cũng gom nhóm các quốc gia/vùng lãnh thổ theo **Region**, tính tổng theo cột **Total Tests** rồi lấy kết quả để điền.

In [15]:
# Gôm nhóm theo region rồi tỉnh tổng theo cột Total Tests
total_tests_pop = first_day_df.groupby('Region')['Total Tests'].sum()
total_tests_pop['All'] = total_tests_pop.sum() # World

# Định nghĩa hàm trả Total Tests khi truyền vào region
def region_total_tests(total_tests_pop, region):
    return total_tests_pop[region]

# Điền giá trị thiếu ở cột Population cho các dòng tổng hợp
first_day_df.loc[0:7, 'Total Tests'] = first_day_df.head(7).apply(lambda row: region_total_tests(total_tests_pop, row['Region']), axis=1)

first_day_df.head(10)

Country,Other  Total Cases  New Cases  Total Deaths  New Deaths  \
0            Asia    112080501   521505.0     1333376.0      1766.0   
1   North America     94158432    57642.0     1384321.0       979.0   
2   South America     53321618    91557.0     1248687.0       984.0   
3          Europe    150719217   660784.0     1681452.0      2030.0   
4          Africa     11442875     8624.0      247689.0       194.0   
5         Oceania      3284142    20201.0        7457.0        49.0   
7           World    425007506  1360313.0     5902997.0      6002.0   
8           China       107707      195.0        4636.0         0.0   
9             USA     80117484    28647.0      959863.0       431.0   
10          India     42838524    16051.0      512141.0       206.0   

    Total Recovered  New Recovered  Active Cases  Serious, Critical  \
0       104076728.0       498347.0     6670397.0            30563.0   
1        63866758.0       164163.0    28907353.0            18121.0   
2        46304681.0       148298.0     5768250.0            13954.0   
3       122959179.0      1165527.0    26078586.0            16595.0   
4        10454264.0        15330.0      740922.0             1984.0   
5         2997343.0        21041.0      279342.0              263.0   
7       350659659.0      2012706.0    68444850.0            81480.0   
8          101449.0           55.0        1622.0                7.0   
9        51675064.0       129104.0    27482557.0            11668.0   
10       42124284.0        37901.0      202099.0             8944.0   

    Tot Cases/1M pop  Deaths/1M pop   Total Tests  Tests/1M pop    Population  \
0                NaN            NaN  1.907488e+09           NaN  4.664923e+09   
1                NaN            NaN  1.050910e+09           NaN  5.966738e+08   
2                NaN            NaN  2.078058e+08           NaN  4.365839e+08   
3                NaN            NaN  2.462184e+09           NaN  7.483700e+08   
4                NaN            NaN  9.517482e+07           NaN  1.393144e+09   
5                NaN            NaN  7.029054e+07           NaN  4.337959e+07   
7            54525.0          757.3  5.793853e+09           NaN  7.883074e+09   
8               75.0            3.0  1.600000e+08      111163.0  1.439324e+09   
9           239741.0         2872.0  9.387964e+08     2809224.0  3.341836e+08   
10           30550.0          365.0  7.601463e+08      542096.0  1.402236e+09   

               Region  
0                Asia  
1       North America  
2       South America  
3              Europe  
4              Africa  
5   Australia/Oceania  
7                 All  
8                Asia  
9       North America  
10               Asia

### Điền các cột có dạng Value/1M pop

Ta có 3 cột theo dạng trên 1 triệu dân số: **Tot Cases/1M pop, Deaths/1M pop** và **Tests/1M pop**. Để điền giá trị thiếu cho các cột này ta đơn giản lấy cột tương ứng chia cho cột **Population** rồi nhân cho 1 triệu.

In [16]:
def col_on_1millpop(col, pop):
    if pop == 0: # Không có dân số không thể tính toán
        return 0
    return col/pop*10**6

first_day_df['Tot Cases/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Cases'], row['Population'])
                                                             if pd.isnull(row['Tot Cases/1M pop']) else row['Tot Cases/1M pop'],
                                                     axis=1)

first_day_df['Deaths/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Deaths'], row['Population'])
                                                             if pd.isnull(row['Deaths/1M pop']) else row['Deaths/1M pop'],
                                                     axis=1)

first_day_df['Tests/1M pop'] = first_day_df.apply(lambda row: col_on_1millpop(row['Total Tests'], row['Population'])
                                                             if pd.isnull(row['Tests/1M pop']) else row['Tests/1M pop'],
                                                     axis=1)

### Điền Region cho 2 chiếc du thuyền

Phần Region của 2 chiếc du thuyền này ta sẽ điền chuỗi rỗng.

In [18]:
two_cruises = first_day_df['Country,Other'].isin(['Diamond Princess', 'MS Zaandam'])
first_day_df.loc[two_cruises, 'Region'] = ''

### Kiểm tra lại còn cột nào thiếu dữ liệu không?

In [19]:
first_day_df.isnull().any()

Country,Other        False
Total Cases          False
New Cases            False
Total Deaths         False
New Deaths           False
Total Recovered      False
New Recovered        False
Active Cases         False
Serious, Critical    False
Tot Cases/1M pop     False
Deaths/1M pop        False
Total Tests          False
Tests/1M pop         False
Population           False
Region               False
dtype: bool

## Thêm 